### PR TITLE
RHPAM-2319: Stunner - Users don't understand how to create lines between nodes

### PR DESCRIPTION
**Pull request by romartin**
Hey @hasys 

Here is the intial PR for https://issues.jboss.org/browse/RHPAM-2319

See comments in https://github.com/kiegroup/kie-wb-common/pull/3012

It depends on:
https://github.com/kiegroup/kie-wb-common/pull/3012
https://github.com/kiegroup/lienzo-tests/pull/78

Thanks!

### CODE REVIEW

File: src/main/java/com/ait/lienzo/client/core/shape/toolbox/items/DecoratedItem.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public interface DecoratedItem<T extends DecoratedItem> extends Item<T> {
+
+    boolean isVisible();
+
+    T decorate(DecoratorItem<?> decorator);
+
+    T tooltip(TooltipItem<?> tooltip);
+
+    // TODO: Refactor by Runnable arg?
+    T onMouseEnter(NodeMouseEnterHandler handler);
```
[the remark this answers is not in the document]
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/toolbox/items/DecoratedItem.java]: yep, it was just a TODO for thinkging on it, but it does not need any change, that's fine using the handler as argument, removing the TODO

File: src/main/java/com/ait/lienzo/client/core/shape/toolbox/items/DecoratedItem.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    // TODO: Refactor by Runnable arg?
```
[the remark this answers is not in the document]
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/toolbox/items/DecoratedItem.java]: yep, it was just a TODO for thinkging on it, but it does not need any change, that's fine using the handler as argument, removing the TODO

File: src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresConnectorProxy.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        this.connectorDestroyer = connectorDestroyer;
+    }
+
+    @Override
+    public void start(final double x,
+                      final double y) {
+        startLocation = new Point2D(x, y);
+        enable();
+        getConnectorControl().showControlPoints();
+        // TODO: Fix or ensure tests use the following order: move primitive -> control.onMoveStart
```
[the remark this answers is not in the document]
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresConnectorProxy.java]: yep, will ensure tests are ok here

File: src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresDragProxy.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        registrations[MOVE] = node.addNodeMouseMoveHandler(new NodeMouseMoveHandler() {
+            @Override
+            public void onNodeMouseMove(NodeMouseMoveEvent moveEvent) {
+                move(moveEvent);
+            }
+        });
+    }
+
+    // TODO: Improve this.
```
[the remark this answers is not in the document]
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresDragProxy.java]: Refactored! Removed this code ,was just temporal... did in the right now, everything via handlers

File: src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresShapeProxy.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    public void start(final double x,
+                      final double y) {
+        final Point2D location = new Point2D(x, y);
+        enable();
+        startLocation = location.copy();
+        if (null != shape.getParent()) {
+            final Point2D parentLocation = shape.getParent().getComputedLocation();
+            startLocation.minus(parentLocation);
+        }
+        // TODO: Fix or ensure tests use the following order: move primitive -> align&distro -> control.onMoveStart
```
[the remark this answers is not in the document]
Comment (romartin) [reply src/main/java/com/ait/lienzo/client/core/shape/wires/proxy/WiresShapeProxy.java]: sure, will ensure it

**Comment by romartin** [comment]
Hey @jomarko 
Thansk for the reivew! Yeah the PR it's quite bit but in fact it's because I moved ALL the code related to the toolbox and its capabilitires to lienzo-core, it was previoulsy on stunner-lienzo-extensions, so now it's much easier to test and fix. So most of the code from your comments it's not really new, it has been moved from stunner, anyway good catrches, I'll take a look at your comments and provide some new commit with the fixes, just waiting for @hasys review too first.

thanks a lot guys, I know it's a big change, but mostly it's a refactor... I moved all toolbox stuff into lienzo-core, so in fact the new stuff is also the classes related to Shape/Connector/Node/Edge proxies... built on top, and some other refactors, but in general most of the stuff is because of refactoring code from one modulje to another... 

Anyway plz lemme know if you need help or whatever for the review! glad to help... I think this code is much much better than the one before, and also the results are better, so I'm jsut open to discussions and feedback, I just want to make our tool better too xD

Thanks!

**Comment by jomarko** [comment]
@romartin I definitely agree the results is better, it seems faster/smoother now.
Unfortunately, I can not asses much the code improvements as I do not do daily with stunner code. But I believe if you claim it is, then it is. The only thing I noticed was heavy using of some magic constants like 0, -1, 1 what is kind of hard to understand, maybe enums/constants would help me, but it can be personal preference, take this comment as an optional.

**Comment by jomarko** [comment]
Manual Review note 01
https://issues.redhat.com/browse/RHPAM-2640

**Comment by jomarko** [comment]
Manual review note 02
No alerts/information are shown to the user when he tries create connection between nodes, where it is not allowed. Only the target node is highlighted with red.

Is this intended? Due to PR size I am not able answer on my own.

**Comment by jomarko** [comment]
@romartin I am done with review, let me know how to proceed. Will we address comments above separately? Should I approve?

**Comment by romartin** [comment]
@jomarko thanks - I'm workng on some fixes and will also check yours. Will let you know

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
@mbiarnes @mareknovotny Hey guys, can you please increase the builds timeout for this module? Also please check [lienzo-tests](https://github.com/kiegroup/lienzo-tests) module too. See [last build error](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/lienzo-core-pullrequests/16/console) - timeout is only 4m, we should increase it please.
Thanks!

**Comment by mareknovotny** [comment]
well, @romartin the timeout on all jobs are by default 180 mins. But we have flexible timeout if possible and this flexible value is counted by Jenkins from multiple several last successful builds, so if there is 4 mins it means that last builds took approx. under 4 mins. If there is some change which increase the build time, we should rerun to try to reset the flexible timeout. Even if it increase is significant the question is why.

**Comment by mareknovotny** [comment]
Jenkins please retest this

**Pull request merged**